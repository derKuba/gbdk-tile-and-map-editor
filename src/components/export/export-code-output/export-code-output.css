
.codeblock {
  background-color: #2d2d2d;
  color: white;
  padding: 15px;
}

.un{
  color: #F012BE;
  font-style: italic;
}

.char{
  color: #F012BE;
  font-style: italic;
}

.clazz{
  color: #3D9970;
}

.arr{
  color: white;
}


.intro {
  display: block;
  color: #c0c0c0;
  font-size: 12px;
  font-style: italic;
  margin-bottom: 15px;
}
.code {
  margin: 20px 0 0 20px;
}

.buttonLine {
  display:flex;
  justify-content: space-between;
}

button {

  border: none;
  
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 16px;
}

button.primary {
  background-color: #4CAF50;
  color: white;
}

button.secondary {
  background-color: #c0c0c0;
  color: black;
}