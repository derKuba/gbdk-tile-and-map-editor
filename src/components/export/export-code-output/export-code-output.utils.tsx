export const createCTemplatesWeb = convertedLines => {
    const linesWithNewLines = convertedLines.join().replace(/(?:[^,]+,){8}/g, '$&<br/>');

    return {
        main: `<span class="intro">// main.c <br/> </span><span class="un">unsigned</span> <span class="char">char</span> <span class="clazz">TileTable </span><span class="arr">[ ]</span> = { <br/> <div class="code">${linesWithNewLines}</div>  <br/> };`,
        header:
            '<span class="intro">// main.h <br/></span> <span class="un">#define SpriteBank 0</span><br/><br/> <span class="un">extern unsigned char</span> <span class="clazz">TileTable </span><span class="arr">[ ]</span>;',
    };
};

export const createCTemplatesFile = convertedLines => {
    return {
        main: `unsigned char Sprites[] = { \n ${convertedLines} \n }`,
        header: '#define SpriteBank 0 \n extern unsigned char Sprites [];',
    };
};
