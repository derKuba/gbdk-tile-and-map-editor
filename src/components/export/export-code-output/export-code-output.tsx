import { Component, h, Prop } from '@stencil/core';
import { convertColorCodeMatrixToHEXLine } from '../../../utils/convertColourCodeMatrixToHex';

import { createCTemplatesWeb, createCTemplatesFile } from './export-code-output.utils';

@Component({
    tag: 'mvjp-export-code-output',
    styleUrl: 'export-code-output.css',
    shadow: false,
})
export class ExportCodeOutput {
    /**
     *  matrix to export
     */
    @Prop() fields: number[][][];

    /**
     *  navigate through overlay
     */
    @Prop() navigateTo: Function;

    private handleClick = () => {
        this.navigateTo();
    };

    private convertArrayToGbdkCompatibelString = (selectedFields: number[][][]) => {
        return selectedFields.map((line: number[][]) => convertColorCodeMatrixToHEXLine(line));
    };

    private createDownload = (content, filename) => {
        const contentType = 'application/octet-stream';
        const a = document.createElement('a');
        const blob = new Blob([content], { type: contentType });
        a.href = window.URL.createObjectURL(blob);
        a.download = filename;
        a.click();
    };

    private downLoadFiles = cTemplatesFile => {
        this.createDownload(cTemplatesFile.main, 'Sprites.c');
        this.createDownload(cTemplatesFile.header, 'Sprites.h');
    };

    render() {
        const convertedLines = this.convertArrayToGbdkCompatibelString(this.fields);
        const cTemplates = createCTemplatesWeb(convertedLines);
        const cTemplatesFile = createCTemplatesFile(convertedLines);

        return (
            <div>
                <h1>Code Output</h1>

                <div class="codeblock">
                    <div innerHTML={cTemplates.main}></div>
                </div>

                <hr />

                <div class="codeblock">
                    <div innerHTML={cTemplates.header}></div>
                </div>

                <hr />

                <div class="buttonLine">
                    <button onClick={this.handleClick} class="secondary">
                        {' '}
                        Back{' '}
                    </button>
                    <button onClick={() => this.downLoadFiles(cTemplatesFile)} class="primary">
                        {' '}
                        Save as File{' '}
                    </button>
                </div>
            </div>
        );
    }
}
