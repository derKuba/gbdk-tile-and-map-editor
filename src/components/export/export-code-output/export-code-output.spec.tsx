import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { ExportCodeOutput } from './export-code-output';

const selectedField = [
    [
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 0, 0, 0, 0, 0, 0],
        [2, 2, 2, 0, 0, 0, 0, 0],
    ],
    [
        [0, 0, 0, 2, 0, 0, 0, 0],
        [0, 0, 2, 2, 2, 0, 0, 0],
        [0, 0, 2, 0, 2, 0, 0, 0],
        [0, 0, 2, 0, 2, 0, 0, 0],
        [0, 0, 2, 2, 2, 0, 0, 0],
        [0, 0, 0, 2, 0, 0, 0, 0],
        [0, 0, 0, 0, 2, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
    ],
];

describe('mvjp-export-code-output', () => {
    it('builds', () => {
        expect(new ExportCodeOutput()).toBeTruthy();
    });

    it('matches snapshot', async () => {
        const page = await newSpecPage({
            components: [ExportCodeOutput],
            template: () => {
                return <mvjp-export-code-output navigateTo={() => {}} fields={selectedField} />;
            },
        });

        expect(page.root).toMatchSnapshot();
    });

    it('should test navigate to', async () => {
        const navigateTo = jest.fn();

        const page = await newSpecPage({
            components: [ExportCodeOutput],
            template: () => {
                return <mvjp-export-code-output navigateTo={navigateTo} fields={selectedField} />;
            },
        });

        page.root.querySelector('button.secondary').click();
        await page.waitForChanges();
        expect(navigateTo).toBeCalled();
    });

    it('should test navigate to', async () => {
        const createURLMock = jest.fn();
        global.Blob = function Blob() {};
        global.window = {
            URL: {
                createObjectURL: createURLMock,
            },
        };

        const page = await newSpecPage({
            components: [ExportCodeOutput],
            template: () => {
                return <mvjp-export-code-output navigateTo={() => {}} fields={selectedField} />;
            },
        });

        page.root.querySelector('button.primary').click();
        await page.waitForChanges();
        expect(createURLMock).toBeCalled();
    });
});
