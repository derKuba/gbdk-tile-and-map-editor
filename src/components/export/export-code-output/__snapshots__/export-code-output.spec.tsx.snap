// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mvjp-export-code-output matches snapshot 1`] = `
<mvjp-export-code-output>
  <div>
    <h1>
      Code Output
    </h1>
    <div class="codeblock">
      <div>
        <span class="intro">
          // main.c
          <br>
        </span>
        <span class="un">
          unsigned
        </span>
        <span class="char">
          char
        </span>
        <span class="clazz">
          TileTable
        </span>
        <span class="arr">
          [ ]
        </span>
        = {
        <br>
        <div class="code">
          0x00, 0xC0, 0x00, 0xC0, 0x00, 0xC0, 0x00, 0xC0,
          <br>
          0x00, 0xC0, 0x00, 0xC0, 0x00, 0xC0, 0x00, 0xE0,
          <br>
          0x00, 0x10, 0x00, 0x38, 0x00, 0x28, 0x00, 0x28,
          <br>
          0x00, 0x38, 0x00, 0x10, 0x00, 0x08, 0x00, 0x00
        </div>
        <br>
        };
      </div>
    </div>
    <hr>
    <div class="codeblock">
      <div>
        <span class="intro">
          // main.h
          <br>
        </span>
        <span class="un">
          #define SpriteBank 0
        </span>
        <br>
        <br>
        <span class="un">
          extern unsigned char
        </span>
        <span class="clazz">
          TileTable
        </span>
        <span class="arr">
          [ ]
        </span>
        ;
      </div>
    </div>
    <hr>
    <div class="buttonLine">
      <button class="secondary">
        Back
      </button>
      <button class="primary">
        Save as File
      </button>
    </div>
  </div>
</mvjp-export-code-output>
`;
